pkg/refactor/events/events.go

```diff
@@ -0,0 +1,107 @@
+package events
+
+import (
+	"context"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/uuid"
+)
+
+// wraps the cloud-event and adds contextual information.
+type Event struct {
+	Event      *cloudevents.Event
+	Namespace  uuid.UUID
+	ReceivedAt time.Time // marks when the events received by the web-API or created via internal logic.
+}
+
+// Persists and gets events.
+type EventHistoryStore interface {
+	// adds at least one and optionally multiple events to the storage.
+	// returns the events that where successfully appended
+	Append(ctx context.Context, event []*Event) ([]*Event, []error)
+	GetByID(ctx context.Context, id uuid.UUID) (*Event, error)
+	// the result will be sorted by the AcceptedAt value.
+	// pass 0 for limit or offset to get all events.
+	// The total row count is also returned for pagination.
+	// keyValues MUST be passed as key value pairs.
+	// passed keyValues will be used filter results.
+	// supported keys are created_before, created_after,
+	// received_before, received_after, event_contains, type_contains.
+	Get(ctx context.Context, limit, offset int, namespace uuid.UUID, keyValues ...string) ([]*Event, int, error)
+	GetAll(ctx context.Context) ([]*Event, error)
+	// deletes events that are older then the given timestamp.
+	DeleteOld(ctx context.Context, sinceWhen time.Time) error
+}
+
+// Helps query the proper event-listeners for a namespace and event-type.
+type EventTopicsStore interface {
+	// topic SHOULD be a compound of namespaceID and the eventType like this: "uuid-eventType"
+	Append(ctx context.Context, namespaceID, eventListenerID uuid.UUID, topic string) error
+	GetListeners(ctx context.Context, topic string) ([]*EventListener, error)
+	Delete(ctx context.Context, eventListenerID uuid.UUID) error
+}
+
+// represents a listener for one or multiple events with specific types.
+type EventListener struct {
+	ID                          uuid.UUID
+	CreatedAt                   time.Time
+	UpdatedAt                   time.Time
+	Deleted                     bool        // set true to remove the subscription.
+	NamespaceID                 uuid.UUID   // the namespace to which the listener belongs.
+	ListeningForEventTypes      []string    // the types of the event the listener is waiting for to be triggered.
+	ReceivedEventsForAndTrigger []*Event    // events already received for the EventsAnd trigger.
+	LifespanOfReceivedEvents    int         // set 0 to omit the value.
+	TriggerType                 TriggerType // set true for EventsAnd.
+	Trigger                     TriggerInfo // hold the information to decide what to do if the listener has satisfied.
+}
+
+type TriggerInfo struct {
```

Comment:
I think we should flatten the `TriggerInfo` into the `EventListener`. I don't think there's any obvious reason why the user of the package should want or care to separate the two. Also, if we flatten it then we can use foreign keys with delete cascades on workflows and instances.

Reply:
Are you sure? Are you using gORM at this stage? I would have thought it works better with a flatter struct.